decidim-proposals/app/commands/decidim/proposals/admin/update_proposal_scope.rb

```diff
@@ -0,0 +1,75 @@
+# frozen_string_literal: true
+
+module Decidim
+  module Proposals
+    module Admin
+      #  A command with all the business logic when an admin batch updates proposals scope.
+      class UpdateProposalScope < Rectify::Command
+        include TranslatableAttributes
+        # Public: Initializes the command.
+        #
+        # scope_id - the scope id to update
+        # proposal_ids - the proposals ids to update.
+        def initialize(scope_id, proposal_ids)
+          @scope = Decidim::Scope.find_by id: scope_id
+          @proposal_ids = proposal_ids
+          @response = { scope_name: "", successful: [], errored: [] }
+        end
+
+        # Executes the command. Broadcasts these events:
+        #
+        # - :update_proposals_scope - when everything is ok, returns @response.
+        # - :invalid_scope - if the scope is blank.
+        # - :invalid_proposal_ids - if the proposal_ids is blank.
+        #
+        # Returns @response hash:
+        #
+        # - :scope_name - the translated_name of the scope assigned
+        # - :successful - Array of names of the updated proposals
+        # - :errored - Array of names of the proposals not updated because they already had the scope assigned
+        def call
+          return broadcast(:invalid_scope) if @scope.blank?
+          return broadcast(:invalid_proposal_ids) if @proposal_ids.blank?
+
+          update_proposals_scope
+
+          broadcast(:update_proposals_scope, @response)
+        end
+
+        private
+
+        attr_reader :scope, :proposal_ids
+
+        def update_proposals_scope
+          @response[:scope_name] = translated_attribute(scope.name, scope.organization)
+          Proposal.where(id: proposal_ids).find_each do |proposal|
+            if scope == proposal.scope
+              @response[:errored] << proposal.title
+            else
+              transaction do
```

Comment:
I'm not sure that this transaction declaration has any effect, because in the case `notify_author` executes then the notification event runs in an async thread. But let's leave here it just in case.

Reply:
So refactoring, and moving the `notify_author` inside `update_proposal_scope` be safer?

such as:
```rb
def update_proposal_scope(proposal)
  proposal.transaction do
    proposal.update!(scope: scope)
    notify_author proposal if proposal.coauthorships.any?
  end
end
```